src/sentry/incidents/logic.py

```diff
@@ -329,6 +329,29 @@ def bulk_build_incident_query_params(incidents, start=None, end=None, prewindow=
     return query_args_list
 
 
+def calculate_incident_rollup(incident):
+    """
+    This function tries to calculate a good rollup for displaying an incidents
+    stat's on a graph. It prioritized a rollup at the rule's time_window,
+    to be consistent with the builder and more easily understand when/where
+    thresholds are passed. This works for most cases, except for incidents
+    with a long duration and small time window (Results are more than MAX_DATA_POINTS - bad)
+    or a short duration and long time window (Very few to no data points - bad).
+    In those cases, we just choose a rollup that results in a reasonable graph.
+    Note: time_window is in minutes, rollup is in seconds.
+    """
+    alert_rule = incident.alert_rule
+    if alert_rule:
+        # When we persist a rule's time_window, switch to using the persisted data.
+        rollup = alert_rule.time_window * 60
+        while ((incident.duration.total_seconds()) / rollup) > MAX_DATA_POINTS:
+            rollup = rollup * 2
+    else:
+        rollup = max(int(incident.duration.total_seconds() / 200), 1)
```

Comment:
Could `total_seconds / 200` still go over MAX_DATA_POINTS? Like if a customer leaves an incident running for a long time and we get the timeseries results for the entire incident?